cmd/untar.go

```diff
@@ -0,0 +1,156 @@
+/*
+ * MinIO Cloud Storage, (C) 2021 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/bzip2"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/klauspost/compress/s2"
+	"github.com/klauspost/compress/zstd"
+	gzip "github.com/klauspost/pgzip"
+	"github.com/pierrec/lz4"
+)
+
+func detect(r *bufio.Reader) format {
+	z, err := r.Peek(4)
+	if err != nil {
+		return formatUnknown
+	}
+	for _, f := range magicHeaders {
+		if bytes.Equal(f.header, z[:len(f.header)]) {
+			return f.f
+		}
+	}
+	return formatUnknown
+}
+
+//go:generate stringer -type=format -trimprefix=format $GOFILE
+type format int
+
+const (
+	formatUnknown format = iota
+	formatGzip
+	formatZstd
+	formatLZ4
+	formatS2
+	formatBZ2
+)
+
+var magicHeaders = []struct {
+	header []byte
+	f      format
+}{
+	{
+		header: []byte{0x1f, 0x8b, 8},
+		f:      formatGzip,
+	},
+	{
+		// Zstd default header.
+		header: []byte{0x28, 0xb5, 0x2f, 0xfd},
+		f:      formatZstd,
+	},
+	{
+		// Zstd skippable frame header.
+		header: []byte{0x2a, 0x4d, 0x18},
+		f:      formatZstd,
+	},
+	{
+		// LZ4
+		header: []byte{0x4, 0x22, 0x4d, 0x18},
+		f:      formatLZ4,
+	},
+	{
+		// Snappy/S2 stream
+		header: []byte{0xff, 0x06, 0x00, 0x00},
+		f:      formatS2,
+	},
+	{
+		header: []byte{0x42, 0x5a},
```

Comment:
```suggestion
		header: []byte{0x42, 0x5a, 'h'},
```
The first characters of BZ2 is `BZ`, so we might get some false detections on TAR files which starts with the file name. Adding the only useful value `h` (huffmann) decreases that likelyhood a lot.